### PR TITLE
Support for unconditional training

### PR DESCRIPTION
**Pull request by nihirv**
This PR is for discussing how best to re-enable support for training models without a text prompt.

Currently, training a model using this repo is coupled with requiring an embedder. When training unconditional models, the embedder is not necessary.

In the initial version of this PR, the code which causes the coupling has been commented out. Hopefully this opens a discussion on how best to add support for unconditional training back into the repo.

I've used the lsun churches dataset as an application for this PR. The relevant config file is: `configs/latent-diffusion/lsun_churches-ldm-kl-8.yaml`

The key changes to note:
1. `main.py` requires a `personalization_config` in a .yaml file. The personalization_config seems to be where we pass language specific parameters (lines 713-725 in dev // 720-732 in my PR)
2. Multiple lines in `ddpm.py` have been commented out. The lines which have been commented out is the conditional language coupling mentioned above.

To re-support unconditional training, 2 things need to happen:
1. `main.py` needs to do its assignments to `config.model.params.personalization_config` only if `personalization_config` is present in the config yaml file
2. The commented out code in `ddpm.py` should only run if conditional language coupling has been specified by the user of the system

My proposal is to check for the existence of `personalization_config` in the .yaml file. If it exists, the code can run as is (in main or dev). If the key doesn't exist, then we can use that to infer that we're not in the conditional language setting and set a flag in `ddpm.py` to not run the language-specific code.

One thing to mention... I eventually want to add support for using an image as a prompt instead of language. If you would like this feature in this branch, then these changes will likely need to be revisited if/when I start working on that

### CODE REVIEW

**Comment by nihirv** [comment]
By the way - the model is currently training. But I have no idea if it'll produce useful results or not for a little while longer (15k steps in and I'm still just seeing noise)

**Comment by nihirv** [comment]
If someone could provide an example of a train/loss they've trained from this repo that'd be valuable. So far the changed code seems to be giving this which seems undesirable (I'll post an update here after a couple thousand iterations have completed):

image

**Comment by lstein** [comment]
I really appreciate your taking this on. Model training is an important feature that has been largely neglected in this repo. Since this PR is primarily for discussion, I'm going to mark it draft until it reaches a more mature point.

**Comment by nihirv** [comment]
Thanks. I'm looking forward to contribute more :).

**Comment by nihirv** [comment]
I just trained another model, this time with more frequent logs.

The model seems to be learning properly until around step 3650. After this step, it goes back to generating pure noise:

Step 3250:
image

Step 3750:
image

Loss curve:
image

**Comment by lstein** [comment]
Any progress on tracking down the scheduler issue?

**Comment by nihirv** [comment]
Apologies - I only have 1 GPU I can train this on so experiment speed is a bit limited.

But no - I haven't been able to debug why this is happening.

I've taken a look at how the scheduler changes the LR and everything there seems to be fine/make sense. I've trained some models without a scheduler but also using the LR that causes this instability. These models seem to train fine.

None of the other config files actually use a scheduler (even the lsun bedrooms file). I wonder if this was left there by mistake..

I'm going to remove the scheduler from this config file and implement the changes to I suggested in the OP

File: ldm/models/diffusion/ddpm.py [hunk 1]
```diff
@@ -665,26 +665,29 @@ def __init__(
...
+        #     param.requires_grad = False
 
-        self.model.eval()
-        self.model.train = disabled_train
-        for param in self.model.parameters():
-            param.requires_grad = False
+        # self.model.eval()
+        # self.model.train = disabled_train
+        # for param in self.model.parameters():
+        #     param.requires_grad = False
```
Comment (nihirv) [inline ldm/models/diffusion/ddpm.py:675]: Needs discussing - This seems to always be setting the model to eval mode?

**Comment by nihirv** [comment]
I've continued to play around around with the scheduler. If you change warmup steps in the config to any non-zero number you'll be able to see this issue quicker. For example, changing warmup steps to 10 will show the instable loss after 10 steps. I've also tried using an l2 loss instead of l1 but this makes no difference.

I'm baffled by this. But removing the scheduler from the config seems to lead the model to train.

Anyway. This PR is ready for review now. I can also add to the README instructions on how to train the model for this dataset

File: ldm/models/diffusion/ddpm.py [hunk 2]
```diff
@@ -748,10 +751,11 @@ def register_schedule(
...
         model = instantiate_from_config(config)
-        self.first_stage_model = model.eval()
-        self.first_stage_model.train = disabled_train
-        for param in self.first_stage_model.parameters():
-            param.requires_grad = False
+        self.first_stage_model = model
+        # self.first_stage_model = model.eval()
+        # self.first_stage_model.train = disabled_train
+        # for param in self.first_stage_model.parameters():
+        #     param.requires_grad = False
```
Comment (nihirv) [inline ldm/models/diffusion/ddpm.py:758]: as above

File: ldm/models/diffusion/ddpm.py [hunk 3]
```diff
@@ -766,10 +770,11 @@ def instantiate_cond_stage(self, config):
...
                 model = instantiate_from_config(config)
-                self.cond_stage_model = model.eval()
-                self.cond_stage_model.train = disabled_train
-                for param in self.cond_stage_model.parameters():
-                    param.requires_grad = False
+                self.cond_stage_model = model
+                # self.cond_stage_model = model.eval()
+                # self.cond_stage_model.train = disabled_train
+                # for param in self.cond_stage_model.parameters():
+                #     param.requires_grad = False
```
Comment (nihirv) [inline ldm/models/diffusion/ddpm.py:777]: as above

**Comment by lstein** [comment]
Is this PR still a draft, or is it ready for testing and review?

**Comment by lstein** [comment]
@nihirv There hasn't been any activity on this PR in 9 days and it is still marked as draft. I'm going to close it now and will happily reopen it if you let me know.

**Comment by nihirv** [comment]
Hi @lstein. Apologies - I seem to have missed the notification that you sent me last week. 

PR is ready to review and can update it as needed
